subgraphs/aave-forks/src/mapping.ts

```diff
@@ -372,707 +513,584 @@ export function _handleReserveDataUpdated(
-  liquidate.liquidatee = borrower.toHexString();
-  liquidate.market = market.id;
-  liquidate.hash = event.transaction.hash.toHexString();
-  liquidate.nonce = event.transaction.nonce;
-  liquidate.logIndex = event.logIndex.toI32();
-  liquidate.asset = debtAsset.id;
-  liquidate.amount = amount;
-  liquidate.amountUSD = amount
-    .toBigDecimal()
-    .div(exponentToBigDecimal(inputToken!.decimals))
-    .times(market.inputTokenPriceUSD);
-  liquidate.profitUSD = liquidate.amountUSD.minus(
-    debtToCover
-      .toBigDecimal()
-      .div(exponentToBigDecimal(debtAsset.decimals))
-      .times(repayTokenMarket.inputTokenPriceUSD)
-  );
+  liquidate.positions = liquidatedPositions;
   liquidate.save();
+}
 
-  protocol.cumulativeLiquidateUSD = protocol.cumulativeLiquidateUSD.plus(
-    liquidate.amountUSD
-  );
-  protocol.save();
-  market.cumulativeLiquidateUSD = market.cumulativeLiquidateUSD.plus(
-    liquidate.amountUSD
+export function _handleFlashLoan(
+  asset: Address,
+  amount: BigInt,
+  account: Address,
+  procotolData: ProtocolData,
+  event: ethereum.Event,
+  premiumAmount: BigInt,
+  flashloanPremium: _FlashLoanPremium
+): void {
+  const market = getMarketFromToken(asset, procotolData);
+  if (!market) {
+    log.warning("[_handleFlashLoan] market for token {} not found", [
+      asset.toHexString(),
+    ]);
+    return;
+  }
+  const manager = new DataManager(
+    market.id,
+    market.inputToken,
+    event,
+    procotolData
   );
-  market.save();
+  const tokenManager = new TokenManager(asset, event);
+  const amountUSD = tokenManager.getAmountUSD(amount);
+  const premiumUSDTotal = tokenManager.getAmountUSD(premiumAmount);
+  const flashloan = manager.createFlashloan(asset, account, amount, amountUSD);
+  flashloan.feeAmount = premiumAmount;
+  flashloan.feeAmountUSD = premiumUSDTotal;
+  flashloan.save();
+
+  let reserveFactor = market.reserveFactor;
+  if (!reserveFactor) {
+    reserveFactor = BIGDECIMAL_ZERO;
+  }
+  const protocolRevenueShare = reserveFactor;
+  let premiumUSDToProtocol = premiumUSDTotal.times(protocolRevenueShare);
+  let premiumUSDToLP = premiumUSDTotal.minus(premiumUSDToProtocol);
+  let premiumUSDToDeduct = premiumUSDTotal;
+  const premiumRateTotal = flashloanPremium.premiumRateTotal;
+  let premiumRateToProtocol = premiumRateTotal.times(protocolRevenueShare);
+  let premiumRateToLP = premiumRateTotal.minus(premiumRateToProtocol);
+
+  if (flashloanPremium.premiumRateToProtocol.gt(BIGDECIMAL_ZERO)) {
+    // according to https://github.com/aave/aave-v3-core/blob/29ff9b9f89af7cd8255231bc5faf26c3ce0fb7ce/contracts/interfaces/IPool.sol#L634
+    // premiumRateToProtocol is the percentage of premium to protocol
+    premiumUSDToProtocol = premiumUSDTotal
+      .times(flashloanPremium.premiumRateToProtocol)
+      .plus(premiumUSDToProtocol);
+    premiumRateToProtocol = premiumRateTotal
+      .times(flashloanPremium.premiumRateToProtocol)
+      .plus(premiumRateToProtocol);
```

Comment:
I think these don't need to be multiplied by the rate...

Once you get `premiumUSDToProtocol ` and `premiumUSDToLP` that is all you need.

`premiumUSDToDeduct` = `premiumUSDToLP` (https://github.com/aave/aave-v3-core/blob/29ff9b9f89af7cd8255231bc5faf26c3ce0fb7ce/contracts/protocol/libraries/logic/FlashLoanLogic.sol#L236)

And `premiumUSDToProtocol` is protocol rev and `premiumUSDToLP` is supply rev.

Reply:
I did this because I want to be precise in the split of premium between LP and protocol. The flashloan premium to LP accrued to LiquidityIndex is split between LP and protocol based on reserve factor. my calculation here include the protocol's share of flashloan premium to LP as protocol revenue.